### DISCUSSION TITLE
Looking for an example of configuring clab for connecting to SNMP from host VM

### DISCUSSION THREAD

**Discussion by danstadler-pdx** [srl-labs/containerlab#1378] [Q&A] [answered]
Hi - I'm new to most of this but my goal is to use Containerlab to set up a few virtual routers, and then from the host VM (Ubuntu 20 in my case) be able to use tools like snmpwalk, to view the SNMP telemetry on the routers.

(Ultimately my goal is to set up snmp_exporter on the host VM so I can get the SNMP telemetry as Prometheus metrics, and feed them into Grafana. Full disclosure: I am an SE at Grafana Labs, and I'm taking myself on a learning curve with SNMP here.)

I have attached my full topology. A few key snippets:

```
  nodes:
    R1:
      kind: srl
      mgmt-ipv4: 172.100.100.11
      publish:
        - udp/161
        - udp/162
      env:
        # boot delay in seconds
        BOOT_DELAY: 30

    border0:
      kind: border0

```
I have successfully authenticated to border0 and stored a token locally.

When I try to deploy the lab, here is what I get about the publish section:

```
INFO[0000] Creating lab directory: /home/ubuntu/grafana-snmp/clab-ext
INFO[0000] Creating docker network: Name="statics", IPv4Subnet="172.100.100.0/24", IPv6Subnet="", MTU="1500"
INFO[0000] Creating container: "SSH"
INFO[0000] Creating container: "R2"
INFO[0000] Creating container: "R1"
INFO[0001] Creating virtual wire: R1:e1-1 <--> R2:e1-1
INFO[0001] Creating container: "border0"
INFO[0001] Running postdeploy actions for Nokia SR Linux 'R1' node
INFO[0001] Running postdeploy actions for Nokia SR Linux 'R2' node
INFO[0001] Creating border0.com tunnels...
ERRO[0001] failed to run postdeploy task for node border0: border0.com does not support socket type "udp". Supported types are "ssh|tls|http|https"

```

Is there an easy example someone can provide about how to accomplish this goal?  
thanks

[danstadler-pdx-topology-yml.txt](https://github.com/srl-labs/containerlab/files/11472205/danstadler-pdx-topology-yml.txt)

**Comment by hellt** [comment, marked as answer]
Hi @danstadler-pdx 
You don't need to use `publish` to just be able to connect to the nodes running on the same host. Just remove `publish` and `BOOT_DELAY` and `border0` node from your lab and you should be able to go

**Reply by danstadler-pdx** [reply]
thank you for the quick reply! 

I also wonder if I'm just going about this the wrong way; I looked around for some examples and came to this: 
https://github.com/jbemmel/srl-sflow-demo/blob/main/srl-test.yml
https://github.com/jbemmel/srl-sflow-demo/blob/main/leaf1-with-sflow.cfg.json#L294

in other words, use `static-config` to point to a JSON body that opens up ports; also, use a full linux box inside the lab, and maybe run snmp_exporter and Grafana agent there, vs trying to run them on the host VM.

Do you have any opinions/direction on these options, in the context of my use case?

thanks again- 
dan

**Reply by hellt** [reply]
You don't need either of them. Port 161 is enabled by default. So just deploy a node without any extra fields, like here https://github.com/srl-labs/containerlab/blob/main/lab-examples/srl01/srl01.clab.yml, and you should be able to use snmpwalk

**Reply by danstadler-pdx** [reply]
very cool, thank you for the help. FWIW I am really enjoying getting to know containerlab; I feel it's going to be very useful for me as an SE at Grafana Labs.

**Reply by hellt** [reply]
you're welcome @danstadler-pdx! 
We are big fans of grafana and use it a lot in our demos as well. For example - https://github.com/srl-labs/srl-telemetry-lab

**Reply by danstadler-pdx** [reply]
That is awesome and I'm going to dig into the details. I would love to see where exactly in that stack a) SNMP gets converted to Prometheus, and b) how the resulting Prometheus telemetry is making it into the Prometheus back-end. Think I can probably figure this out - but will let you know.

**Reply by danstadler-pdx** [reply]
I get 2 errors during deploy; not sure if this is due to anything I've left behind from other labs I've deployed. 

`ERRO[0001] failed deploy phase for node "prometheus": Error response from daemon: driver failed programming external connectivity on endpoint prometheus (866afb82c38630078cec74e454f28ca85ce0c34d3a8ce346d49bc33ef61324fe): Error starting userland proxy: listen tcp4 0.0.0.0:9090: bind: address already in use`

`ERRO[0004] failed to run postdeploy task for node prometheus: failed to Statfs "": no such file or directory`

Going to start seeing if that means something is broken or not - let me know if you have any thoughts on what's going on. (If it helps: I'm on an Ubuntu 20 EC2 instance, 2XL, and have installed docker and containerlab per standard published steps.)

**Reply by danstadler-pdx** [reply]
the prometheus container seems not to be able to run. Tried this on a completely new Ubuntu EC2, just to sanity check that it's not something left behind from other testing. If you can help solve this it would be great. I am open to an early AM screenshare (I'm in Portland, OR) if that helps speed the conversation. Feel free to connect on LI and we can DM from there: https://www.linkedin.com/in/danstadler/

**Reply by hellt** [reply]
it looks like someone is already occupying 9090 port on the host. Maybe an embedded prometheus? You may want to change the port in the `binds` knob from 9090 to smth else

**Reply by danstadler-pdx** [reply]
yeah that is weird. I have never seen a fresh Ubuntu instance with an unknown process holding onto 9090. 
<img width="1199" alt="Screen Shot 2023-05-15 at 12 27 54 PM" src="https://github.com/srl-labs/containerlab/assets/10214489/15bb8d8d-6a3d-46fe-b06e-e957a3c8781f">

Anyway I fixed that in st.clab.yaml, so now moving on to next steps.

     prometheus:
      kind: linux
      mgmt-ipv4: 172.80.80.42
      image: prom/prometheus:v2.37.8
      binds:
        - configs/prometheus/prometheus.yml:/etc/prometheus/prometheus.yml:ro
      cmd: --config.file=/etc/prometheus/prometheus.yml --web.listen-address=0.0.0.0:9095
      ports:
        - 9095:9095
      group: "10"

**Reply by hellt** [reply]
Likely AWS uses it for metrics collection, over a standard port.

**Reply by danstadler-pdx** [reply]
ah. makes sense.

**Reply by danstadler-pdx** [reply]
I got back to this a bit more today. I am slightly stumped and wondering if you could offer me a cheat sheet set of quick steps to follow.

Here's what I can do no problem: 
- create a topology with 2 Ubuntu 20.04 VMs, and on one of them, configure snmpd, then snmpwalk it from the other Ubuntu VM

I can also add an SRLinux VM to the topology, and update its snmpd.conf so that I can snmpwalk it from itself (localhost). What I can't quite figure out: 
- how to snmpwalk the SRL VM, from the Ubuntu VM

When I try that, I get:
Timeout: No Response from 172.80.80.13

(I can ping to 172.80.80.13 - but I just can't get snmpwalk to connect.

Am I just missing something simple?

thanks thanks thanks
dan

**Reply by danstadler-pdx** [reply]
to close this out: thank you for the great help over on Discord. https://discord.com/channels/854262008589058058/854262008589058061/1108397848514535475

Long story short, I now understand a lot more about the following:

```
enter candidate
set / system snmp community public
set / system snmp network-instance mgmt
set / system snmp network-instance mgmt admin-state enable
commit now
```

as well as the documentation here:
https://documentation.nokia.com/srlinux/23-3/books/system-mgmt/configuring-snmp-system-mgmt.html

**Reply by hellt** [reply]
great to hear it worked out @danstadler-pdx
We are open to collaboration with grafana on networking stuff, so it would be nice to see you using containerlab for internal tests/dev
